### PR TITLE
refactor(x/ecocredit): move empty string validation into utils.go

### PR DESCRIPTION
**Pull request by wgwz**
## Description

Ref: https://github.com/regen-network/regen-ledger/issues/1245

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/basket/features/msg_put.feature [hunk 1]
```diff
@@ -95,7 +95,7 @@ Feature: MsgPut
     }
     """
     When the message is validated
-    Then expect the error "invalid batch denom: expected format A00-000-00000000-00000000-000: parse error: invalid request"
+    Then expect the error "invalid batch denom: expected format A00-000-00000000-00000000-000: invalid request"
```
Comment (ryanchristo) [inline x/ecocredit/basket/features/msg_put.feature:98]: Here's an example of what might be the preferred format but the deleted `parse error` might actually make sense to keep as the error in the utility function and the error returned within basic message validation would wrap the returned error with an `invalid request` error.

File: x/ecocredit/basket/msg_put.go [hunk 2]
```diff
@@ -39,12 +39,8 @@ func (m MsgPut) ValidateBasic() error {
 
 	if len(m.Credits) > 0 {
 		for _, credit := range m.Credits {
-			if len(credit.BatchDenom) == 0 {
-				return sdkerrors.ErrInvalidRequest.Wrap("credit batch denom cannot be empty")
-			}
-
 			if err := core.ValidateBatchDenom(credit.BatchDenom); err != nil {
-				return sdkerrors.ErrInvalidRequest.Wrap(err.Error())
+				return err
```
Comment (ryanchristo) [inline x/ecocredit/basket/msg_put.go:43]: Here's an example of where we should wrap the error with `sdkerrors.ErrInvalidRequest`.

The error returned from the utility function should not already be wrapped with the `sdkerrors.ErrInvalidRequest` as the utility function may be used outside the context of a request and the parse error might b the preferred error to return in the utility function, which is then wrapped here with `sdkerrors.ErrInvalidRequest`.

File: x/ecocredit/core/features/msg_bridge_receive.feature [hunk 3]
```diff
@@ -129,7 +129,7 @@ Feature: MsgBridgeReceive
     }
     """
     When the message is validated
-    Then expect the error "invalid jurisdiction: foo, expected format <country-code>[-<region-code>[ <postal-code>]]: parse error: invalid request"
+    Then expect the error "invalid jurisdiction: foo, expected format <country-code>[-<region-code>[ <postal-code>]]: invalid request: invalid request"
```
Comment (ryanchristo) [inline x/ecocredit/core/features/msg_bridge_receive.feature:132]: In line with my comments above, I think the previous error is the desired error message here, which would include the parse error from the utility function and the invalid request error from basic message validation.

**Review by ryanchristo — COMMENTED** [review]
Thanks for the contribution! Looks like a good start.

The issue was not very specific so my apologies. Although the parse error that was previously being used is defined in the ecocredit module, it is registered as an sdk error.

```
ErrParseFailure        = sdkerrors.Register(ModuleName, 2, "parse error")
```

I think this error still makes sense when returning an error for an invalid format, and this error would be preferred over using `sdkerrors.ErrInvalidRequest` within the utility functions given the utility functions may be used outside the context of a request.

I've added another note to an issue proposing guidelines for better use of sdk errors, which is also being used to track examples (https://github.com/regen-network/regen-ledger/issues/1147#issuecomment-1199793577).

**Comment by wgwz** [comment]
hey @ryanchristo 👋 i made another pass at this and pushed a new commit. i have still only done the refactoring work for `x/ecocredit/core/utils.go`. it should be closer now, LMK what you think, thanks!

File: x/ecocredit/core/features/msg_retire.feature [hunk 4]
```diff
@@ -127,4 +127,21 @@ Feature: MsgRetire
...
+        {
+          "batch_denom": "C01-001-20200101-20210101-001",
+          "amount": "100"
+        }
+      ],
+      "jurisdiction": ""
+    }
+    """
+    When the message is validated
+    Then expect the error "jurisdiction cannot be empty, expected format <country-code>[-<region-code>[ <postal-code>]]: parse error: invalid request"
```
Comment (ryanchristo) [inline x/ecocredit/core/features/msg_retire.feature:147]: This is a duplicate of `Scenario: an error is returned if jurisdiction is empty`. If jurisdiction is not provided, the zero-value is an empty string, which is why the error is the same here as it is above.

```suggestion
```

**Review by ryanchristo — COMMENTED** [review]
This looks good to me. The added scenario is a duplicate but otherwise I think this is ready if you want to open a separate pull request to address basket and marketplace utils.

File: x/ecocredit/core/features/msg_retire.feature (hunk 4 above)
Comment (wgwz) [reply x/ecocredit/core/features/msg_retire.feature]: Ah ok, I see. I have removed this.

**Comment by wgwz** [comment]
@ryanchristo I do like the idea of merging this and then addressing the other locations in separate PRs. because i may not have the time this week to handle the others. should we open new separate issues? or favor keeping #1245 open and adding a todo list?

**Review by ryanchristo — APPROVED** [review]
Looks good to me. Thanks for the contribution!

**Review by technicallyty — APPROVED** [review]
👍🏻

**Pull request merged**